notes/version-3-proposal.rst

```diff
@@ -0,0 +1,326 @@
+methods).
+
+**2. Single level of JSON loading**
+
+This one is very easy to fix.  Instead of just assigning each top-level key to
+a local attribute, just JSON parse the entire ``metadata`` section and store
+it.  This will parse (and thus, JSON validate) the attributes.  They can then
+be lazily accessed via ``__getattribute__()`` if they exist in the
+``_metadata``.
+
+**3. .client, .api properties on every lxd object**
+
+The end result will be that every attribute on an lxd object will have a direct
+mapping to the (new) ``_metadata`` private attribute on the object.  i.e. the
+``class Model`` will have a private ``_metadata`` member that is the
+``metadata`` returned by API calls to lxd.
+
+The ``.save()``, ``.dirty()``, ``.sync()``, ``.rollback()``, ``.delete()`` and
+``.marshal()`` methods will be removed completely, and new functions will be
+added to the top level pylxd module to access their functionality.  Therefore,
+saving an lxd object will then take the form:
+
+.. code-block:: python
+
+   import pylxd
+
+   ...
+   pylxd.save(some_object)
+
+The concept is that the objects themselves become just data entities, but will
+a little bit of behaviour around read-only attributes, and 'manager' attributes
+for 'chained' objects (e.g. profiles on a container).
+
+This ``client`` and ``api`` properties will be renamed to ``_client`` and
+``_api`` to make them 'private' and thus, for internal library use; i.e.
+signalling that the user of the library is not supposed to use these methods on
+the objects.  If possible, they may be removed entirely, although it's not
+clear how dependent objects could be automatically fetched without having at
+least the ``_client`` available on each object.
+
+Note, that currently, the ``.api`` property is set for each object to resolve
+the API path and instead having an ``_url`` property which returns the API
+endpoint url fragment for that container.  Then the ``save(...)``,
+``sync(...)`` can use that as a data string, rather than as a convoluted client
+accessing property which is hard to follow when reading the code.
+
+**4.  The object system is only partially done**
+
+e.g. Container objects have profiles, but they just return strings.  i.e. to
+access a profile, currently:
+
+.. code-block:: python
+
+   import pylxd
+
+   c = pylxd.Client()
+   container = c.containers.get('test')
+   profile = c.profiles.get(container.profiles[0])
+
+This would be much more natural if it could be accessed as:
+
+.. code-block:: python
+
+   import pylxd
+
+   c = pylxd.Client()
+   profile = c.containers['test'].profiles[0]
+
+and the ``profile`` object would be a pylxd ``Model()`` object, where the
+``used_by`` attribute would be a dictionary like object which had 'test' as a
+key, and the test container as a value. e.g.
+
+.. code-block:: python
+
+   container = c.profiles['default'].used_by
+   <... Manager for containers ...>
+
+i.e. the ``used_by`` property would return a manager to access profiles.
```

Comment:
I believe that this manager would access containers?

Reply:
And, yes, this is true too.